cmds/echo/echo.go

```diff
@@ -17,19 +17,85 @@ import (
 	"strings"
 )
 
-var nonewline = flag.Bool("n", false, "suppress newline")
+type flags struct {
+	noNewline, interpretEscapes bool
+}
+
+func escapeString(s string) (string, error) {
+	if len(s) < 1 {
+		return "", nil
+	}
+
+	s = strings.Split(s, "\\c")[0]
+	s = strings.Replace(s, "\\0", "\\", -1)
+
+	// Quote the string and scan it through %q to interpret backslash escapes
+	s = fmt.Sprintf("\"%s\"", s)
+	_, err := fmt.Sscanf(s, "%q", &s)
+	if err != nil {
+		return "", err
+	}
+
+	return s, nil
+}
 
-func echo(w io.Writer, s ...string) error {
-	_, err := fmt.Fprintf(w, "%s", strings.Join(s, " "))
+func echo(f flags, w io.Writer, s ...string) error {
+	var err error
+	line := strings.Join(s, " ")
+	if f.interpretEscapes {
+		line, err = escapeString(line)
+		if err != nil {
+			return err
+		}
 
-	if !*nonewline {
-		fmt.Fprint(w, "\n")
 	}
 
+	format := "%s"
+	if !f.noNewline {
+		format += "\n"
+	}
+	_, err = fmt.Fprintf(w, format, line)
+
 	return err
 }
 
+func init() {
+	defUsage := flag.Usage
+	flag.Usage = func() {
+		defUsage()
+		fmt.Println(`
+  If -e is in effect, the following sequences are recognized:
+    \\     backslash
+    \a     alert (BEL)
+    \b     backspace
+    \c     produce no further output
+    \e     escape
+    \f     form feed
+    \n     new line
+    \r     carriage return
+    \t     horizontal tab
+    \v     vertical tab
+    \0NNN  byte with octal value NNN (1 to 3 digits)
+    \xHH   byte with hexadecimal value HH (1 to 2 digits)`)
+	}
+}
+
 func main() {
+	var (
+		f flags
+		E bool
+	)
+	flag.BoolVar(&f.noNewline, "n", false, "suppress newline")
```

Comment:
Here's a case for moving these out to file scope: if these were placed right near the top of the file, then it would have been much easier to see that the usage string in the comment was wrong. Further, f and E are in some sense global and ought to be in file scope anyway? Just a thought.

overall this is fine.